src/Identity/v3/Service.php

```diff
@@ -31,7 +31,15 @@ public function authenticate(array $options): array
     {
         $authOptions = array_intersect_key($options, $this->api->postTokens()['params']);
 
-        $token = $this->generateToken($authOptions);
+        if (!empty($options['cachedToken'])) {
+            $token = $this->generateTokenFromCache($options['cachedToken']);
+
+            if ($token->hasExpired()) {
+                throw new \RuntimeException(sprintf('Cached token has expired. You may need to re-generate a new token.'));
```

Comment:
nit: since we're using `sprintf` how about output the expired date?

Reply:
Good idea.